src/Analyzers/Core/Analyzers/ForEachCast/AbstractForEachCastDiagnosticAnalyzer.cs

```diff
@@ -81,12 +80,39 @@ protected override void InitializeWorker(AnalysisContext context)
                 return;
 
             var (conversion, collectionElementType) = GetForEachInfo(semanticModel, node);
+            if (collectionElementType is null)
+                return;
 
             // Don't bother checking conversions that are problematic for other reasons.  The user will already have a
             // compiler error telling them the foreach is in error.
             if (!conversion.Exists)
                 return;
 
+            // Consider:
+            // public class C : IEnumerable<Match>
+            // {
+            //     public IEnumerator GetEnumerator() => null; // compiler picks this for the foreach loop.
+            //
+            //     IEnumerator<Match> IEnumerable<Match>.GetEnumerator() => null; // compiler doesn't use this.
+            // }
+
+            // This collection have GetEnumerator method that returns non-strongly-typed IEnumerator, but also implements strongly-typed IEnumerable<T> explicitly.
+            // In this case, the compiler considers the non-strongly-typed GetEnumerator and adds explicit cast for `foreach (Match m in new C())`.
+            // This cast can fail if the collection is badly implemented such that the strongly-typed and non-strongly-typed GetEnumerator implemetations return different types.
+            // So, we want to consider the collectionElementType as System.Object (that's what the compiler returns).
+            // However, we trust CorLib, so we adjust collectionElementType for the case above to be `Match` instead of object if the type comes from CorLib.
+            if (collectionElementType.SpecialType == SpecialType.System_Object &&
+                collectionType.ContainingAssembly is { } collectionTypeContainingAssembly &&
+                collectionTypeContainingAssembly.Identity.PublicKey.SequenceEqual(context.Compilation.GetSpecialType(SpecialType.System_Object).ContainingAssembly?.Identity.PublicKey))
```

Comment:
Regex and System.Object come from different assemblies, but the assemblies have the same public key. So I relied on that to know whether `collectionType` is something built-in or not.

Reply:
I wasn't sure why we enforce that these come from the same assembly. Does it matter if the collection is built-in or not?

Reply:
I see technically that someone could write code that intentionally fails this (e.g. your sharplab example) but I think its probably rare enough that I wouldn't bother doing this check.  If not we'll hear feedback otherwise

Reply:
> I see technically that someone could write code that intentionally fails this (e.g. your sharplab example) but I think its probably rare enough that I wouldn't bother doing this check. If not we'll hear feedback otherwise

Yeah that was my intention to guard against this rare case. Going to revert this part now.